src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,998 @@
+		int i;
+		for (i = (start >> 3) + 1; i < (end >> 3); ++i)
+			if (bitmap[i] != 0xFF)
+				return 0;
+	}
+
+	return 1;
+}
+
+/**
+ * This function clears a continuous range of bits in the bitmap.
+ * The range if from start to end inclusive.
+ */
+static inline void
+clear_bitmap_range(uint8_t *bitmap, uint64_t start, uint64_t end)
+{
+	uint8_t mask = (0xFF >> (start % 8));
+
+	mask = ~mask;
+	if (end >> 3 == start >> 3) {
+		mask |= (0xFF >> ((end%8)+1));
+		bitmap[(start >> 3)] &= mask;
+		return;
+	}
+
+	bitmap[(start >> 3)] &= mask;
+
+	mask = (0xFF >> ((end % 8) + 1));
+	bitmap[(end >> 3)] &= mask;
+
+	if (end >> 3 > (start >> 3) + 1)
+		memset(&(bitmap[(start>>3)+1]), 0, (end>>3) - ((start>>3)+1));
+}
+
+/**
+ * This function returns the number of non-leaf domains in the pool map
+ * It's used to determine the size of the arrays for storing the bookkeeping
+ * bitmaps
+ *
+ * \param[in]	dom	The pool map used for this placement map.
+ *
+ * \return		The number of non-leaf nodes in the pool, not including
+ *			the root.
+ */
+uint64_t get_dom_cnt(struct pool_domain *dom)
+{
+	uint64_t count = 0;
+
+	if (dom->do_children != NULL) {
+		int i;
+		count += dom->do_child_nr;
+
+		for (i = 0; i < dom->do_child_nr; ++i)
+			count += get_dom_cnt(&dom->do_children[i]);
+	}
+
+	return count;
+}
+
+/**
+ * Jump Consistent Hash Algorithm that provides a bucket location
+ * for the given key. This algorithm hashes a minimal (1/n) number
+ * of keys to a new bucket when extending the number of buckets.
+ *
+ * \param[in]   key             A unique key representing the object that
+ *                              will be placed in the bucket.
+ * \param[in]   num_buckets     The total number of buckets the hashing
+ *                              algorithm can choose from.
+ *
+ * \return                      Returns an index ranging from 0 to
+ *                              num_buckets representing the bucket
+ *                              the given key hashes to.
+ */
+static inline uint32_t
+jump_consistent_hash(uint64_t key, uint32_t num_buckets)
+{
+        int64_t z = -1;
+        int64_t y = 0;
+
+        while (y < num_buckets) {
+                z = y;
+                key = key * 2862933555777941757ULL + 1;
+                y = (z + 1) * ((double)(1LL << 31) /
+                               ((double)((key >> 33) + 1)));
+        }
+        return z;
+}
+
+/**
+ * Assembly instruction for fast 4-byte CRC computation on Intel X86
+ * Copied from BSD-licensed Data Plane Development Kit (DPDK) rte_hash_crc.h
+ *
+ * \param[in] data              Primary input to hash function
+ * \param[in] init_value        Acts like a seed to the CRC, useful to get
+ *                              different results with the same data
+ */
+static inline uint32_t
+crc32c_sse42_u32(uint32_t data, uint32_t init_val)
+{
+        __asm__ volatile(
+                "crc32l %[data], %[init_val];"
+                : [init_val] "+r"(init_val)
```

Comment:
(style)  code indent should use tabs where possible
(style)  space prohibited before open square bracket '['